feat(internal/db): Storage policies implementation

hugoghx opened the pull request:
This commit implements part of the storage policy's database model, focusing specifically on the operational tables for this feature.

hashicorp-cla commented:
[![CLA assistant check](https://cla.hashicorp.com/pull/badge/not_signed)](https://cla.hashicorp.com/hashicorp/boundary?pullRequest=3816)


Thank you for your submission! We require that all contributors sign our Contributor License Agreement ("CLA") before we can accept the contribution. [Read and sign the agreement](https://cla.hashicorp.com/hashicorp/boundary?pullRequest=3816)

[Learn more about why HashiCorp requires a CLA and what the CLA includes](https://www.hashicorp.com/cla)


---
**Hugo** seems not to be a GitHub user.
You need a GitHub account to be able to sign the CLA. If you already have a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).

<sub>Have you signed the CLA already but the status is still pending? [Recheck it](https://cla.hashicorp.com/check/hashicorp/boundary?pullRequest=3816).</sub>

hugoghx commented on internal/db/schema/migrations/oss/postgres/78/01_storage_policies.up.sql:
```diff
@@ -0,0 +1,160 @@
+'scope_id being inserted is either global or the same org as the storage policy'
+'we want to associate';
+
+-- TBD: Given that storage_policy_id is not immutable, should there also be a
+-- 'before update' trigger with the same procedure?
```
TBD

mgaffney commented on internal/db/schema/migrations/oss/postgres/78/01_storage_policies.up.sql:
```diff
@@ -0,0 +1,160 @@
+comment on table policy is
+'policy contains entries that represent an abstract policy';
+
+create table storage_policy (
```
```suggestion
create table policy_storage_policy (
```
Storage policy is in the `policy` package since it is a type of policy.

mgaffney commented on internal/db/schema/migrations/oss/postgres/78/01_storage_policies.up.sql:
```diff
@@ -0,0 +1,160 @@
+create trigger storage_policy_scope_id_valid before insert on storage_policy
+  for each row execute procedure storage_policy_scope_id_valid();
+
+create or replace function insert_policy_subtype() returns trigger
```
```suggestion
create function insert_policy_subtype() returns trigger
```

mgaffney commented on internal/db/schema/migrations/oss/postgres/78/01_storage_policies.up.sql:
```diff
@@ -0,0 +1,160 @@
+create trigger insert_policy_subtype before insert on storage_policy
+  for each row execute procedure insert_policy_subtype();
+
+create or replace function delete_policy_subtype() returns trigger
```
```suggestion
create function delete_policy_subtype() returns trigger
```

mgaffney commented on internal/db/schema/migrations/oss/postgres/78/01_storage_policies.up.sql:
```diff
@@ -0,0 +1,160 @@
+create trigger delete_policy_subtype after delete on storage_policy
+  for each row execute procedure delete_policy_subtype();
+
+create table scope_storage_policy_override_behavior_enm (
```
```suggestion
create table policy_storage_policy_override_behavior_enm (
```

mgaffney commented on internal/db/schema/migrations/oss/postgres/78/01_storage_policies.up.sql:
```diff
@@ -0,0 +1,160 @@
+
+-- TBD: There's no triggers on storage_policy that would populate this table.
+-- Will this happen in code?
+create table scope_storage_policy (
```
```suggestion
create table policy_assigned_storage_policy (
```

mgaffney commented on internal/db/schema/migrations/oss/postgres/78/01_storage_policies.up.sql:
```diff
@@ -0,0 +1,160 @@
+create trigger immutable_columns before update on scope_storage_policy
+  for each row execute procedure immutable_columns('scope_id');
+
+create or replace function scope_storage_policy_scope_id_valid() returns trigger
```
```suggestion
create function scope_storage_policy_scope_id_valid() returns trigger
```

hugoghx commented:
@mgaffney I think I addressed all your feedback with [0c0eb0f](https://github.com/hashicorp/boundary/pull/3816/commits/0c0eb0fea03bd1ebd0c87ebd7d604fd817a0766f). Let me know if I missed anything!

hugoghx commented on internal/db/schema/migrations/oss/postgres/78/01_storage_policies.up.sql:
```diff
@@ -0,0 +1,160 @@
+create table policy_storage_policy (
+  public_id wt_public_id primary key,
+  scope_id wt_scope_id not null,
+  retention_days integer not null,
```
TBD: `bigint` instead?
